File: leet_easy_c++/leet_112/leet_112a.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root) return false;
        bool l, r;
        if(!(root->right) && !(root->left) &&  sum - (root->val)==0) return true;
        l = hasPathSum(root->left, sum - (root->val));
        r = hasPathSum(root->right, sum - (root->val));
        return max(l,r);
    }
};

File: leet_easy_python/leet_112/leet_112a.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        if not root: return False
        v = sum - root.val
        if not root.left and not root.right and v == 0: return True
        return self.hasPathSum(root.left, v) or self.hasPathSum(root.right, v)
        
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """